In [ ]:
# recursion assignment

#Q.1 Can you explain the logic and working of the Tower of Hanoi algorithm by writing a Java program?
How does the recursion work, and how are the movements of disks between rods accomplished?

In [2]:

#Certainly! Here's an example of the Tower of Hanoi problem using Python, solving it for three disks:
def move_disks(n, source, auxiliary, destination):
    if n == 1:
        print(f"Move disk 1 from rod {source} to rod {destination}")
        return
    
    move_disks(n - 1, source, destination, auxiliary)
    print(f"Move disk {n} from rod {source} to rod {destination}")
    move_disks(n - 1, auxiliary, source, destination)

def tower_of_hanoi(num_of_disks):
    move_disks(num_of_disks, 'A', 'B', 'C')

# Solving Tower of Hanoi for 3 disks
tower_of_hanoi(3)


Move disk 1 from rod A to rod C
Move disk 2 from rod A to rod B
Move disk 1 from rod C to rod B
Move disk 3 from rod A to rod C
Move disk 1 from rod B to rod A
Move disk 2 from rod B to rod C
Move disk 1 from rod A to rod C


Explanation:

The move_disks function is responsible for moving disks.
When n is 1, it's the base case. It moves the smallest disk directly from the source rod to the destination rod.
For n > 1, the function follows these steps:
It moves n-1 disks from the source rod to the auxiliary rod using the destination rod as the auxiliary.
It then moves the n-th disk (the largest one) from the source rod to the destination rod.
Finally, it recursively moves the n-1 disks from the auxiliary rod to the destination rod using the source rod as the auxiliary.
This process continues recursively until all disks are moved from the source rod to the destination rod, adhering to the rules of the Tower of Hanoi problem (moving one disk at a time and never placing a larger disk on top of a smaller one).

The movements of disks between rods are managed by the recursive calls within the move_disks function. Each recursive call handles a smaller subproblem of moving a subset of disks, eventually solving the Tower of Hanoi problem for the specified number of disks.







In [3]:
#Q.2 Given two strings word1 and word2, return the minimum number of operations required to convert word1
#to word2.

Certainly! This problem can be solved using dynamic programming, specifically the concept of edit distance. The minimum number of operations required to convert one string to another can be calculated using a dynamic programming table.

Here's a Python solution to find the minimum number of operations (edit distance) between two words:

In [4]:
def minDistance(word1, word2):
    m, n = len(word1), len(word2)

    # Initializing a DP table
    dp = [[0] * (n + 1) for _ in range(m + 1)]

    # Filling the DP table
    for i in range(m + 1):
        for j in range(n + 1):
            if i == 0:
                dp[i][j] = j
            elif j == 0:
                dp[i][j] = i
            elif word1[i - 1] == word2[j - 1]:
                dp[i][j] = dp[i - 1][j - 1]
            else:
                dp[i][j] = 1 + min(dp[i - 1][j - 1], dp[i][j - 1], dp[i - 1][j])

    return dp[m][n]

# Test cases
word1_1, word2_1 = "horse", "ros"
word1_2, word2_2 = "intention", "execution"

print(minDistance(word1_1, word2_1))  # Output: 3
print(minDistance(word1_2, word2_2))  # Output: 5


3
5


Explanation:

The minDistance function takes two input strings, word1 and word2, and computes the minimum number of operations required to convert word1 into word2.
It initializes a DP table dp with dimensions (m+1) x (n+1), where m and n are the lengths of word1 and word2 respectively.
It fills the DP table using a bottom-up approach:
The base cases are when either of the words is an empty string. In these cases, the number of operations needed is the length of the other word.
For non-empty strings, it checks if the characters at the corresponding positions in word1 and word2 match. If they match, the value in the DP table is the same as the value at the previous position.
If the characters don't match, it takes the minimum of three operations:
Replacing a character (dp[i - 1][j - 1])
Inserting a character (dp[i][j - 1])
Deleting a character (dp[i - 1][j])
Finally, it returns dp[m][n], which contains the minimum number of operations needed to convert word1 into word2.
This algorithm computes the minimum edit distance between two words efficiently using dynamic programming.

In [5]:
#Q. 3 Print the max value of the array [ 13, 1, -3, 22, 5].

Certainly! To find the maximum value of the array [13, 1, -3, 22, 5] in Python, you can use the max() function. Here's an example:

In [8]:
arr = [13, 1, -3, 22, 5]
max_value = max(arr)
print("The maximum value in the array is:", max_value)


The maximum value in the array is: 22


This code will output:

The maximum value in the array is: 22

The max() function in Python returns the maximum value present in the given iterable, in this case, the list [13, 1, -3, 22, 5].

In [9]:
#Q.4 Find the sum of the values of the array [92, 23, 15, -20, 10].


Sure, to find the sum of the values in the array [92, 23, 15, -20, 10] in Python, you can use the sum() function. Here's how you can do it:

In [10]:
arr = [92, 23, 15, -20, 10]
sum_of_values = sum(arr)
print("The sum of the values in the array is:", sum_of_values)


The sum of the values in the array is: 120


This code will output:

The sum of the values in the array is: 120

The sum() function in Python calculates the sum of all the elements in the given iterable, in this case, the list [92, 23, 15, -20, 10], and returns the total sum.

In [11]:
#Q.5 Given a number n. Print if it is an armstrong number or not.An armstrong number is a number if the sum
#of every digit in that number raised to the power of total digits in that number is equal to the number.


Sure, here's a Python function that checks whether a given number is an Armstrong number or not:

In [12]:
def is_armstrong_number(n):
    # Convert the number to a string to find its length (number of digits)
    num_str = str(n)
    num_digits = len(num_str)
    
    # Calculate the sum of each digit raised to the power of total digits
    sum_of_digits = sum(int(digit) ** num_digits for digit in num_str)
    
    # Check if the sum is equal to the original number
    if sum_of_digits == n:
        return "Yes"
    else:
        return "No"

# Test cases
input1 = 153
input2 = 134

print("Input 1:", is_armstrong_number(input1))  # Output: Yes
print("Input 2:", is_armstrong_number(input2))  # Output: No


Input 1: Yes
Input 2: No


Explanation:

The is_armstrong_number function takes an input number n.
It calculates the number of digits in n by converting it to a string and finding its length (num_digits).
It then computes the sum of each digit raised to the power of num_digits using a generator expression and the sum() function.
Finally, it checks if the calculated sum is equal to the original number and returns "Yes" if they are equal, otherwise "No".
This function checks whether a given number is an Armstrong number based on the provided conditions.